x-pack/plugin/ilm/src/main/java/org/elasticsearch/xpack/ilm/PolicyStepsRegistry.java

```diff
@@ -60,9 +63,11 @@ public class PolicyStepsRegistry {
     // keeps track of existing policies in the cluster state
     private final SortedMap<String, LifecyclePolicyMetadata> lifecyclePolicyMap;
     // keeps track of what the first step in a policy is, the key is policy name
-    private final Map<String, Step> firstStepMap;
+    private final Map<VersionedPolicyKey, Step> firstStepMap;
     // keeps track of a mapping from policy/step-name to respective Step, the key is policy name
-    private final Map<String, Map<Step.StepKey, Step>> stepMap;
+    private final Map<VersionedPolicyKey, Map<Step.StepKey, Step>> stepMap;
```

Comment:
So stepMap is a bit tricky.

Given we don't allow removing actions across versions, so a later version will always contain all the actions in a previous version, and the fact that the stepMap is solely used in boolean stepExists(final String policy, final Step.StepKey stepKey) (note we do NOT return the Step that's in the stepMap), one could argue we don't technically need it versioned - it could hold only the latest actions version of the policy.

And that's true. The latest actions order will have all the step KEYs that exist in the previous versions. The value, the Step associated with a step KEY, could be different (if actions are reordered, the last step KEY in an action will point to different STEPs across actions order) however we never expose the Step.

So we could get away with not versioning the stepMap however it felt error prone to me. If we decide to expose the Step associated with a step key to the outside world we'd only ever expose the latest version Step (it'd be a very subtle bug). Also it feels wrong to me to have a steps registry that's not away of all the versions of the steps.